compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentGeneratorTest.kt

```diff
@@ -0,0 +1,399 @@
+      assertThat(messages).contains("Source0.kt: (9, 3)")
+      assertThat(messages).contains("A factory must be an interface or an abstract class.")
+    }
+
+    compile(
+      """
+        package com.squareup.test
+  
+        import com.squareup.anvil.annotations.ContributesSubcomponent
+        import com.squareup.anvil.annotations.ContributesSubcomponent.Factory
+        import com.squareup.anvil.annotations.ContributesTo
+  
+        @ContributesSubcomponent(Any::class, parentScope = Unit::class)
+        interface SubcomponentInterface {
+          @Factory
+          class ComponentFactory {
+            fun createComponent(): SubcomponentInterface = throw NotImplementedError()
+          }
+        }
+      """
+    ) {
+      assertThat(exitCode).isError()
+      assertThat(messages).contains("Source0.kt: (9, 3)")
+      assertThat(messages).contains("A factory must be an interface or an abstract class.")
+    }
+  }
+
+  @Test
+  fun `a factory must have a single abstract method returning the subcomponent - no function`() {
+    compile(
+      """
+        package com.squareup.test
+  
+        import com.squareup.anvil.annotations.ContributesSubcomponent
+        import com.squareup.anvil.annotations.ContributesSubcomponent.Factory
+        import com.squareup.anvil.annotations.ContributesTo
+  
+        @ContributesSubcomponent(Any::class, parentScope = Unit::class)
+        interface SubcomponentInterface {
+          @Factory
+          interface ComponentFactory
+        }
+      """
+    ) {
+      assertThat(exitCode).isError()
+      assertThat(messages).contains("Source0.kt: (9, 3)")
+      assertThat(messages).contains(
+        "A factory must have exactly one abstract function returning the subcomponent " +
+          "com.squareup.test.SubcomponentInterface."
+      )
+    }
+  }
+
+  @Test
+  fun `a factory must have a single abstract method returning the subcomponent - two functions`() {
+    compile(
+      """
+        package com.squareup.test
+  
+        import com.squareup.anvil.annotations.ContributesSubcomponent
+        import com.squareup.anvil.annotations.ContributesSubcomponent.Factory
+        import com.squareup.anvil.annotations.ContributesTo
+  
+        @ContributesSubcomponent(Any::class, parentScope = Unit::class)
+        interface SubcomponentInterface {
+          @Factory
+          interface ComponentFactory {
+            fun createComponent1(): SubcomponentInterface
+            fun createComponent2(): SubcomponentInterface
+          }
+        }
+      """
+    ) {
+      assertThat(exitCode).isError()
+      assertThat(messages).contains("Source0.kt: (9, 3)")
+      assertThat(messages).contains(
+        "A factory must have exactly one abstract function returning the subcomponent " +
+          "com.squareup.test.SubcomponentInterface."
+      )
+    }
+  }
+
+  @Test
+  fun `a factory must have a single abstract method returning the subcomponent - non-abstract function`() {
```

Comment:
I think we're missing one or two tests here?
- Must supply `@ContributesSubcomponent.Factory`  not `@Subcomponent.Factory`
- Fail on `@Subcomponent.Builder` <- I kept on forgetting and kept on using the builder and wondered why I got dagger errors.

Reply:
Good suggestion: https://github.com/square/anvil/pull/435